lib/src/main/java/com/cloudbees/groovy/cps/impl/BlockScopeEnv.java

```diff
@@ -12,16 +14,34 @@
  */
 // TODO: should be package local once all the impls move into this class
 public class BlockScopeEnv extends ProxyEnv {
-    private final Map<String,Object> locals = new HashMap<String, Object>();
-    private Map<String, Class> types = new HashMap<String, Class>();
+    private Map<String,Object> locals;
+    private Map<String, Class> types;
 
     public BlockScopeEnv(Env parent) {
+        this(parent, 0);
+    }
+
+    public BlockScopeEnv(Env parent, int localsSize) {
         super(parent);
+        if (localsSize <= 0) {
+            locals = Collections.emptyMap();
+            types = Collections.emptyMap();
+        } else {
+            locals = Maps.newHashMapWithExpectedSize(localsSize);
+            types = Maps.newHashMapWithExpectedSize(localsSize);
+        }
     }
 
     public void declareVariable(Class type, String name) {
+        if (locals == Collections.EMPTY_MAP) {
+            this.locals = new HashMap<String, Object>(2);
+        }
         locals.put(name, null);
-        getTypes().put(name, type);
+
+        if (types == null || types == Collections.EMPTY_MAP) {
```

Comment:
Beware that the Java Platform does not guarantee that `Collections.emptyMap() == Collections.EMPTY_MAP`, though this is true in Oracle JRE 8 at least. The Javadoc says only

> **Implementation Note:** Implementations of this method need not create a separate `Map` object for each call. Using this method is likely to have comparable cost to using the like-named field. (Unlike this method, the field does not provide type safety.)